src/java.base/share/classes/java/lang/ThreadLocal.java

```diff
@@ -807,4 +798,43 @@ private void expungeStaleEntries() {
             }
         }
     }
+
+
+    /**
+     * Reads the value of the jdk.traceVirtualThreadLocals property to determine if
+     * a stack trace should be printed when a virtual threads sets a thread local.
+     */
+    private static boolean traceVirtualThreadLocals() {
+        String propValue = GetPropertyAction.privilegedGetProperty("jdk.traceVirtualThreadLocals");
+        return (propValue != null)
+                && (propValue.isEmpty() || Boolean.parseBoolean(propValue));
+    }
+
+    /**
+     * Print a stack trace if the current thread is a virtual thread.
+     */
+    static void dumpStackIfVirtualThread() {
+        if (Thread.currentThread() instanceof VirtualThread vthread) {
+            try {
+                var stack = StackWalkerHolder.STACK_WALKER.walk(s ->
+                        s.skip(1)  // skip caller
+                         .collect(Collectors.toList()));
+
+                // switch to carrier thread to avoid recursive use of thread-locals
+                vthread.executeOnCarrierThread(() -> {
+                    System.out.println(vthread);
+                    for (StackWalker.StackFrame frame : stack) {
+                        System.out.format("    %s%n", frame.toStackTraceElement());
+                    }
+                    return null;
+                });
+            } catch (Exception e) {
+                throw new InternalError(e);
```

Comment:
Should inability to log/trace `ThreadLocal` creation or `set` lead to those operations failing? Or would it be OK, if we just ignored this exception that happened when tracing/logging?

Reply:
> Or would it be OK, if we just ignored this exception that happened when tracing/logging?

If, for example, something has changed System.out to be a throwing PrintStream, and this diagnostic option is enabled, then it may fail here. As it's just a diagnostic option then I don't think it's a bit issue. Also this code will change with a future PR that will integrate this will filtering.